### PR TITLE
PDS-69180 - CommitTsLoader is too aggressive when warming its cache

### LINKED ISSUE #2979: Sweep should batch loads from _transactions
Saw this message when sweeping a wide row on a production stack:
```
Re-batching in getLoadWithTsTasksForSingleHost a call to {} for table {} that attempted to multiget {} rows; this may indicate overly-large batching on a higher level. {}
tableRef: "_transactions"
rows: "117147.00"
```
And again with rows = 116123, 7300, and 109687

Sweep attempted to fetch over 100,000 rows from the `_transactions` table. This gets rebatched to a maximum of 5000 rows; we should do some batching ourselves, perhaps `min(16*maxCellTsPairsToExamine, 5000)`.

### PR DESCRIPTION
**Pull request by jeremyk-91**
**Goals (and why)**: 
- fix suspected root cause of internal ticket PDS-69180
- fix #2979 - see more details there

**Implementation Description (bullets)**:
- If Sweep is checking the status of a ton of candidates, split its reads into smaller batches of 5000, and warm the cache one batch at a time instead of all in parallel.

**Concerns (what feedback would you like?)**:
- should 5000 be configurable? Might be worth doing this, though I refrained owing to high priority and also because this is not a regression (existing is already a flat 5000)
- this obviously slows this phase of sweep down, but I think that's what we want to do!

**Where should we start reviewing?**: `CommitTsLoader.java`

**Priority (whenever / two weeks / yesterday)**: yesterday 🔥

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/3245)

### CODE REVIEW

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/CommitTsLoader.java [hunk 1]
```diff
@@ -49,12 +52,22 @@ private CommitTsLoader(TLongLongMap commitTsByStartTs, TransactionService transa
     public static CommitTsLoader create(TransactionService transactionService, TLongSet startTssToWarmingCache) {
         TLongLongMap cache = new TLongLongHashMap();
         if (!startTssToWarmingCache.isEmpty()) {
-            // Ideally TransactionService should work with primitive collections to avoid GC overhead..
-            cache.putAll(transactionService.get(TDecorators.wrap(startTssToWarmingCache)));
+            warmCache(transactionService, startTssToWarmingCache, cache);
```
Comment (sandorw) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/CommitTsLoader.java]: Curious why we don't consider warming the cache async, especially if we may be loading a million pairs during the warming phase. Especially since the default cache size is one million, so loading any more than that is going to end up being useless.

Comment (sandorw) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/CommitTsLoader.java]: (Not necessarily in the scope of this PR - we should resolve the P0 as a higher priority)

Comment (sandorw) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/CommitTsLoader.java]: As @j-baker noted, sweep is also be a bad user of the cache, since it pollutes it with timestamps that are likely to be uncorrelated with the timestamps read by live requests.

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/CommitTsLoaderTest.java [hunk 2]
```diff
@@ -88,4 +95,26 @@ public void loadShouldThrowIfANullIsToBeReturned() throws Exception {
...
+            return timestamps.stream().collect(Collectors.toMap(n -> n, n -> n));
+        }).when(mockTransactionService).get(any());
+
+        TLongSet initialTimestamps = new TLongHashSet();
+        initialTimestamps.addAll(LongStream.range(0, valuesToInsert).boxed().collect(Collectors.toList()));
+
+        CommitTsLoader loader = CommitTsLoader.create(mockTransactionService, initialTimestamps);
+        assertThat(loader.load(valuesToInsert - 1)).isEqualTo(valuesToInsert - 1);
+
+        verify(mockTransactionService, atLeast(2)).get(any());
```
Comment (sandorw) [inline atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/CommitTsLoaderTest.java]: Is this trying to avoid assuming the 5000 batch size? It's already checked explicitly above. I'd prefer to just make the batch size protected and refer to it directly.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/CommitTsLoader.java (hunk 1 above)
Comment (sandorw) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/CommitTsLoader.java]: Checked the above and we _only_ use the cache for sweep, not for actual transactions, where I'd guess it may be more useful...

Comment (sandorw) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/CommitTsLoader.java]: Fun. 

Transactions use a `TimestampCache` which is by default limited to 1 million entries. It's a dumb cache class with no logic for loading timestamps or rolling back transactions.

Sweep uses a different class `CommitTsLoader` instead, which doesn't really act as a cache. Each `CommitTsLoader` only lives as long as it takes to complete a call to `SweepableCellFilter.getCellsToSweep`, so we only benefit from the cache if, if I understand correctly, multiple cells in a single row were written with the same timestamp, which certainly should happen. That said, we are throwing away the cache completely as soon as we move on to the next batch. Background sweep should use a different cache from the transactions, since the timestamps they load are very likely to be different, but we might benefit from combining the classes used and not blowing away the cache so frequently.

Also, `CommitTsLoader` has no eviction and targeted sweep keeps a long lived one around... filed https://github.com/palantir/atlasdb/issues/3246.

Sorry for all the confusion above, it was pretty surprising to me that we have two separate commit timestamp cache implementations :(

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CellLoader.java [hunk 3]
```diff
@@ -140,8 +140,11 @@ void loadWithTs(String kvsMethodName,
             final byte[] col = entry.getKey();
             Collection<Cell> columnCells = entry.getValue();
             if (columnCells.size() > fetchBatchCount) {
```
Comment (gmaretic) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CellLoader.java]: Should this even be configured by this config? I'd err on the side of making it constant. Also make sure it is consistent with the `TS_PER_BATCH` below

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/CommitTsLoader.java [hunk 4]
```diff
@@ -38,6 +39,8 @@
 public final class CommitTsLoader {
     private static final Logger log = LoggerFactory.getLogger(CommitTsLoader.class);
 
+    private static final int TS_PER_BATCH = 5000;
```
Comment (gmaretic) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/CommitTsLoader.java]: See above

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/CommitTsLoader.java (hunk 1 above)
Comment (jeremyk-91) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/CommitTsLoader.java]: Correct, yeah these are two distinct caches 😞

Yep, rather confusingly we also throw away the `CommitTsLoader`. (Technically the cache gives you benefits if your rows are small and rows that are nearby share timestamps too)

I wrote this PR as a fairly quick hack to defuse the P0, but I'd want to unify the caches at some point

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/CommitTsLoaderTest.java (hunk 2 above)
Comment (jeremyk-91) [reply atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/CommitTsLoaderTest.java]: :+1: will change this

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/CommitTsLoader.java [hunk 5]
```diff
@@ -38,6 +39,8 @@
 public final class CommitTsLoader {
```
Comment (gmaretic) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/CommitTsLoader.java]: This class should really just use a LoadingCache internally so that we have eviction, I'm happy to do that in a separate PR

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/CommitTsLoader.java [hunk 6]
```diff
@@ -49,12 +52,22 @@ private CommitTsLoader(TLongLongMap commitTsByStartTs, TransactionService transa
...
         if (!startTssToWarmingCache.isEmpty()) {
-            // Ideally TransactionService should work with primitive collections to avoid GC overhead..
-            cache.putAll(transactionService.get(TDecorators.wrap(startTssToWarmingCache)));
+            warmCache(transactionService, startTssToWarmingCache, cache);
+
         }
         return new CommitTsLoader(cache, transactionService);
     }
 
+    private static void warmCache(
```
Comment (gmaretic) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/CommitTsLoader.java]: I'd like to get rid of this and try to reuse the cache instead. But outside the scope of this PR.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/CommitTsLoader.java (hunk 5 above)
Comment (jeremyk-91) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/CommitTsLoader.java]: Yes. Will be coming in part 2

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CellLoader.java (hunk 3 above)
Comment (jeremyk-91) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CellLoader.java]: We don't have visibility of `CassandraConstants` here. I can shove this into `AtlasDbConstants` maybe - will have a look at doing that
